2_process.R

```diff
@@ -88,72 +90,143 @@ p2_targets_list <- list(
-                       ## Join with PRMS segment table + clean cols. NOTE: there are two PRMS_segid that match same hru_segment at the moment (nrow change) - to resolve. 
-                       left_join(y=p2_PRMS_hru_segment, by = 'hru_segment') %>% select(PRMS_segid,  everything()) %>% ## n = 418
-                       ## Adding Year column
-                       mutate(Year = .y)}
-                       )}
+                     aggregate_proportions_hrus(group_by_segment_colname = hru_segment,
+                                                proportion_col_prefix = 'prop_lcClass',
+                                                hru_area_colname = hru_area,
+                                                new_area_colname = total_PRMS_area) %>% ## n = 416
+                     ## Join with PRMS segment table + clean cols. NOTE: there are two PRMS_segid that match same hru_segment at the moment (nrow change) - to resolve. 
+                     left_join(y=p2_PRMS_hru_segment, by = 'hru_segment') %>% select(PRMS_segid,  everything()) %>% ## n = 418
+                     ## Adding Year column
+                     mutate(Year = .y)}
+                 )
+    }
+  ),
+  
+  ## Produce subset of p1_prms_reach_attr for p2_FORESCE_LC_per_catchment_reclass_tot target
+  tar_target(
+    p2_prms_attribute_df, 
+    p1_prms_reach_attr %>% select(subseg_id,subseg_seg,from_segs,to_seg) %>% 
+      # Update `from_segs` col by splitting the individual segs in a list (can then loop through the list) 
+      mutate(from_segs = stringr::str_split(string = from_segs, pattern = ';', simplify = F))
     ),
   
+  tar_target(
+    p2_FORESCE_LC_per_catchment_reclass_tot,
+    {lapply(p2_FORESCE_LC_per_catchment_reclass_cat, function(x) 
+      p2_prms_attribute_df %>% 
+        rowwise() %>%
+        # Collect all upstream segs per individual seg_id using recursive_fun() (row wise application)
+        mutate(all_from_segs = list(recursive_fun(x = subseg_seg,  df = p2_prms_attribute_df, col1 = 'subseg_seg', col2 = 'from_segs'))) %>%
+        # unest to have new rows for each upstream catchment
+        unnest(all_from_segs, keep_empty = TRUE) %>%
+        # change col type to be able to compute
+        dplyr::mutate(all_from_segs = as.integer(all_from_segs)) %>%
+        # join prop calculations - selected inner join because at the moment, p2_prms_attribute_df has more PRMS_segids than p2_FORESCE_LC_per_catchment_reclass_cat due to outdated catchmetns file
+        inner_join(x, by = c('all_from_segs' = 'hru_segment')) %>%
+        # group by PRMS id 
+        group_by(PRMS_segid, Year) %>% 
+        summarise(
+          # calc total area
+          total_upstream_PRMS_area = sum(total_PRMS_area),
+          # get proportions for the new total area
+          across(starts_with('prop'), ~(sum((.x*total_PRMS_area)/total_upstream_PRMS_area))),
+          .groups = 'drop_last') %>%
+        drop_na()
+      )}
+  ),
+  
```

Comment:
I agree that this is better as a separate function

Reply:
Ended up as a start to split the steps more evenly between the two targets `p2_prms_attribute_df` and `p2_FORESCE_LC_per_catchment_reclass_tot` 

Waffling between setting this as is or into function because of the high never of different columns that are modified here and deciding what to be a function arg vs staying a local variable.